module/zfs/zfs_vnops.c

```diff
@@ -1445,10 +1446,25 @@ zfs_create(struct inode *dip, char *name, vattr_t *vap, int excl,
 		}
 		zfs_mknode(dzp, vap, tx, cr, 0, &zp, &acl_ids);
 
+		error = zfs_link_create(dl, zp, tx, ZNEW);
+		if (error != 0) {
+			/*
+			 * Delete the node created above.
+			 */
+			zp->z_unlinked = 1;
+			drop_nlink(ZTOI(zp));
+			zfs_unlinked_add(zp, tx);
```

Comment:
adding this to the unlinked set seems like a roundabout way of removing it.  Can we just call zfs_znode_delete()?  I know zfs_inactive->zfs_zinactive->zfs_rmnode has some additional logic, but I don't think that it applies in this case.  E.g. this znode can not have an xattrdir at this point (I think).

The same applies to symlink and mkdir.

Reply:
Will do that.